File: 11.1-BinarySearchTree/programs/cpp/program1.cpp
// Binary search tree implementation in C++

#include<bits/stdc++.h>
using namespace std;

struct Node {
    struct Node *lchild;
    int data;
    struct Node *rchild;
}*root = NULL;

void insert(int key) {
    struct Node *t = root;
    struct Node *r = NULL, *p;

    if (root == NULL) {
        p = new Node;
        p->data = key;
        p->lchild = p->rchild = NULL;
        root = p;
        return;
    }

    while (t != NULL) {
        r = t;
        if (key < t->data)
            t = t->lchild;
        else if (key > t->data)
            t = t->rchild;
        else
            return;
    }

    p = new Node;
    p->data = key;
    p->lchild = p->rchild = NULL;

    if (key < r->data) r->lchild = p;
    else r->rchild = p;
}

void inorder(struct Node *p) {
    if (p) {
        inorder(p->lchild);
        cout << p->data << ", " << flush;
        inorder(p->rchild);
    }
}

struct Node *search(int key) {
    struct Node *t = root;

    while (t != NULL) {
        if (key == t->data)
            return t;
        else if (key < t->data)
            t = t->lchild;
        else
            t = t->rchild;
    }
    return NULL;
}

int main() {

    insert(10);
    insert(5);
    insert(20);
    insert(8);
    insert(30);

    inorder(root);
    cout << endl;

    struct Node *temp = search(10);
    if (temp != NULL) {
        cout << temp->data << " is found" << endl;
    } else {
        cout << "Key not found" << endl;
    }

    return 0;
}

File: 11.1-BinarySearchTree/programs/ruby/program1.rb
# Binary search tree implementation in C++


class Node
  attr_accessor :data, :lchild, :rchild

  def initialize(data)
    @data = data
    @lchild = nil
    @rchild = nil
  end
end

class BinarySearchTree
  def initialize
    @root = nil
  end

  def insert(key)
    if @root.nil?
      @root = Node.new(key)
      return
    end

    t = @root
    r = nil

    while t
      r = t
      if key < t.data
        t = t.lchild
      elsif key > t.data
        t = t.rchild
      else
        return
      end
    end

    p = Node.new(key)
    if key < r.data
      r.lchild = p
    else
      r.rchild = p
    end
  end

  def inorder(p)
    if p
      inorder(p.lchild)
      print "#{p.data}, "
      inorder(p.rchild)
    end
  end

  def search(key)
    t = @root

    while t
      if key == t.data
        return t
      elsif key < t.data
        t = t.lchild
      else
        t = t.rchild
      end
    end
    nil
  end

  def display_inorder
    inorder(@root)
    puts
  end
end

bst = BinarySearchTree.new
bst.insert(10)
bst.insert(5)
bst.insert(20)
bst.insert(8)
bst.insert(30)

bst.display_inorder

temp = bst.search(10)
if temp
  puts "#{temp.data} is found"
else
  puts "Key not found"
end
